router/courses.js

```diff
@@ -293,35 +287,27 @@ function  generateListings(data, user, cb) {
         var debug = false;
         if (debug || (user != '' && user != undefined)) {
             var classid = e.courseOfferingId;
-            permission.checkCoursePermission(user, classid, 'Modify', function(error, result) {
-
+            permission.isManagingAllowed(user, classid).then( result => {
```

Comment:
This flow should be changed a bit,
Instead of calling isManagingAllowed and isWatchingAllowed for every course

Also, in this workflow, you're fetching all courses by university Id. This way you aren't getting courses which are public from other universities.


The workflow I'd suggest is,
Get courseId list by allWatchableCourses,
Get another courseId list by allManageableCourses
Fetch course information corresponding to every courseId of allWatchableCourses. (Try doing this in one sql query)
For all courseIds in allManageableCourses mark those course as manageable.

Makes sense?

Reply:
I have updated the workflow.

NOTE: 
Current Workflow would automatically ignore rows that contain NULL values. This is probably related to how the SQL query is written.